Handling default networks (bridge, null, host) with local store

mavenugo opened the pull request:
Pushing this fix to chenchun's branch to make it easy to update 
the PR https://github.com/docker/libnetwork/pull/466.
Please test it out before merging it in your branch.

With the introduction of Local store, we store all the networks and
endpoints created for all the locally scoped objects.
But, the default networks are special. On daemon start, docker will
install the 3 default networks (docker0 bridge, null and host). But,
since we restore the networks from the local store in controller.New()
the actual network creation from Daemon will fail. Hence, we need to
handle that scenario. To add more to it, the user could have changed
the bridge configuration in docker daemon configuration. Hence the
NewNetwork call from the docker daemon must be honored and at the same
time we have to retain the network ID and endpoints stored in the
local-db.

The solution is to introduce a new state called `started` for the
controller & playback the network and endpoint states from the datastore
only when the Start() is called. Between the New() and Start(), the
Daemon can create the special networks which will be adjusted with the
local states & installed by merging the latest configs with the saved
data.

Signed-off-by: Madhu Venugopal <madhu@docker.com>

chenchun commented on store.go:
```diff
@@ -303,7 +344,9 @@ func (c *controller) processNetworkUpdate(nws []*store.KVPair, prune *networkTab
 
 		if err = c.newNetworkFromStore(&n); err != nil {
-			log.Error(err)
+			if _, ok := err.(NetworkNameError); !ok {
```
If it's `NetworkNameError`, how can drivers merge latest configs with configs in the local store? Are you planning these in the following PR?

mavenugo replied:
For local store, this will never happen, because we dont support config updates today (except for the docker0 bridge) and that is the reason why I had to write this PR on the first place.

But, even if we have to support config updates for other user defined network, then this function has to change regardless and we are not changing anything in this change. As you can see, I just hide the log.Error if the error is NetworkNameError. The functionality stays as today.

chenchun commented on controller.go:
```diff
@@ -69,6 +69,8 @@ type NetworkController interface {
 	// ConfigureNetworkDriver applies the passed options to the driver instance for the specified network type
 	ConfigureNetworkDriver(networkType string, options map[string]interface{}) error
```
How can drivers handle `ConfigureNetworkDriver` in the between the `New()` and `Start()`? For bridge driver, if users changed `EnableIPForwarding` from true to false, it seems bridge driver will ignore this?

chenchun commented:
Also I've found a strange problem with boltdb. Have tested with your PR, I found bridge network will be persisted twice if recreating network after restarting.

```
kv := makeBoltDBClient(t)
assert.Nil(t, kv.Put("TestBoldDBStore", []byte("1"), nil))
pair, err := kv.Get("TestBoldDBStore")
assert.Nil(t, err)
assert.Equal(t, string(pair.Value), "1")
assert.Nil(t, kv.Put("TestBoldDBStore", []byte("2"), nil))
pair, err = kv.Get("TestBoldDBStore")
assert.Equal(t, string(pair.Value), "2")
kv.Close()
kv = makeBoltDBClient(t)
pair, err = kv.Get("TestBoldDBStore")
assert.Equal(t, string(pair.Value), "2")
assert.Nil(t, kv.Put("TestBoldDBStore", []byte("3"), nil))
pairs, err := kv.List("TestBoldDBStore")
assert.Nil(t, err)
assert.Equal(t, len(pairs), 1)
assert.Equal(t, string(pair.Value), "3")
```

Boltdb backend fails at the last assert. It's too late today. I will take a look tomorrow.

chenchun commented on controller.go:
```diff
@@ -281,6 +283,21 @@ func (c *controller) NewNetwork(networkType, name string, options ...NetworkOpti
+		// for locally scoped networks
+		n := c.getNetworkByNameFromLocalStore(name)
+		if n != nil {
+			n.generic = network.generic
```
Should also assign endpoints to n? or n.endpoints will be nil.

mavenugo replied:
Apologies on the delay. I think this should be okay, because this is purely for the network creation. Ideally when network is created, the endpoints is infact nil. When the controller moves to the `started` phase, we do read and recreate network and endpoints back into the cache.

chenchun replied:
It's really needed because when we are reading back network from boltdb, it will result in a `NetworkNameError` and `n.endpoints` is still nil. Then we read back endpoints and when we try to add an endpoint to `n.endpoints`, it will result in a nil point panic.

mavenugo replied:
I think you are correct. instead of adding it here, am adding it as part of setupNetwork which does the same think in a more generic way. Will update this PR with the changes.

mavenugo replied:
Pushed a change with the above fix.

chenchun commented:
Tested with bridge network with the following tests
```
func TestLocalRestore(t *testing.T) {
	defer os.Remove(defaultLocalStoreConfig.Client.Address)
	netOptions := []config.Option{}
	netOptions = append(netOptions, config.OptionLocalKVProvider(""))

	ctrl, err := New(netOptions...)
	if err != nil {
		t.Fatalf("Error new controller: %v", err)
	}

	if err := ctrl.ConfigureNetworkDriver("bridge", options.Generic{netlabel.GenericData: options.Generic{}}); err != nil {
		t.Fatalf("Error initializing bridge driver: %v", err)
	}
	_, bipNet, err := net.ParseCIDR("172.18.42.1/16")
	if err != nil {
		t.Fatalf("Erorr %v", err)
	}
	netOption := options.Generic{
		"BridgeName":         "docker0",
		"EnableIPMasquerade": true,
		"EnableICC":          true,
		"AddressIPv4":        bipNet,
	}
	// Initialize default network on "bridge" with the same name
	nw, err := ctrl.NewNetwork("bridge", "bridge", NetworkOptionGeneric(options.Generic{netlabel.GenericData: netOption}))
	if err != nil {
		t.Fatalf("Error creating default \"bridge\" network: %v", err)
	}
	store := ctrl.(*controller).localStore.KVStore()
	if exists, err := store.Exists(datastore.Key(datastore.NetworkKeyPrefix, string(nw.ID()))); !exists || err != nil {
		t.Fatalf("Network key should have been created.")
	}
	ep, err := nw.CreateEndpoint("newendpoint", []EndpointOption{}...)
	if err != nil {
		t.Fatalf("Error creating endpoint: %v", err)
	}
	if exists, err := store.Exists(datastore.Key([]string{datastore.EndpointKeyPrefix, string(nw.ID()), string(ep.ID())}...)); !exists || err != nil {
		t.Fatalf("Endpoint key should have been created. ")
	}
	err = ctrl.Start()
	if err != nil {
		t.Fatalf("Error starting controller: %v", err)
	}

	store.Close()

	ctrl, err = New(netOptions...)
	if err != nil {
		t.Fatalf("Error new controller: %v", err)
	}

	if err := ctrl.ConfigureNetworkDriver("bridge", options.Generic{netlabel.GenericData: options.Generic{}}); err != nil {
		t.Fatalf("Error initializing bridge driver: %v", err)
	}
	nw, err = ctrl.NewNetwork("bridge", "bridge", NetworkOptionGeneric(options.Generic{netlabel.GenericData: netOption}))
	if err != nil {
		t.Fatalf("Error creating default \"bridge\" network: %v", err)
	}

	err = ctrl.Start()
	if err != nil {
		t.Fatalf("Error starting controller: %v", err)
	}

	if len(nw.Endpoints()) != 1 {
		t.Fatalf("num of endpoints should be 1")
	}
}
```

When loading endpoints from local store after restarting, bridge driver will return error  https://github.com/docker/libnetwork/blob/master/drivers/bridge/bridge.go#L860

mavenugo commented:
@chenchun this is a very useful test and it caught another fundamental problem in bridge driver.
I have pushed a new set of changes (including a slight modification to this test).
Since the `started` mode is added just to handle the case of adding default networks and it is NOT intended to have containers active in that phase, I added code to fail any endpoint create during this phase and modified the test to honor this.

But now, the issue is in bridge driver (https://github.com/docker/libnetwork/blob/master/drivers/bridge/bridge.go#L887). The restore is happening properly, but the bridge driver refuses to honor the request due to the above restriction. ping @mrjana @aboch for suggestions on the best way to handle this basic scenario.

@chenchun in the meantime can you please give it a try and give feedback on the existing fix ?
We need to get this in ASAP. multiple PRs are pending on this one :(

chenchun commented:
@mavenugo I will update the original PR very soon

mavenugo commented:
@chenchun if you update the original PR, it will fail due to `https://github.com/docker/libnetwork/blob/master/drivers/bridge/bridge.go#L887` (as I included your test along with my code). If you can find a way to resolve this issue in bridge driver, then we are good to go I think. Essentially we have to make bridge driver honor the endpoint info from libnetwork core.

chenchun commented:
@mavenugo updated PR https://github.com/docker/libnetwork/pull/466. I tested `TestLocalRestore` locally and it's passed. I've also test against docker, everything seems ok.

mavenugo commented:
@chenchun thanks a ton. I gave 1 final comment : https://github.com/docker/libnetwork/pull/466#discussion_r38899573. Can you please address that in your commit ?
